### ISSUE TITLE
Memory Consumption with large datasets

### ISSUE THREAD

**Issue by eric-sciberras** [typesense/typesense#162]
## Description
Hi, we are indexing entries into Typesense and wanted to get an idea if the memory consumption was reasonable for the size and schema.

Currently we have approximately 2.5 Million records indexed and are consuming approximately 1.8 GB of memory.

The schema for the our collection is defined below:

```
{
    'fields': [
      {
        "facet": false,
        "name": "num",
        "optional": false,
        "type": "int32"
      },
      {
        "facet": false,
        "name": "col1",
        "optional": false,
        "type": "string"
      },
      {
        "facet": false,
        "name": "col2",
        "optional": false,
        "type": "string"
      },
      {
        "facet": false,
        "name": "col3",
        "optional": false,
        "type": "string"
      },
      {
        "facet": false,
        "name": "col4",
        "optional": false,
        "type": "string"
      },
      {
        "facet": false,
        "name": "col5",
        "optional": false,
        "type": "string[]"
      },
      {
        "facet": false,
        "name": "col6",
        "optional": false,
        "type": "string[]"
      },
      {
        "facet": false,
        "name": "col7",
        "optional": false,
        "type": "string"
      },
      {
        "facet": false,
        "name": "col8",
        "optional": false,
        "type": "string"
      },
      {
        "facet": false,
        "name": "col9",
        "optional": false,
        "type": "string"
      }
    ],
    'default_sorting_field': 'num'
}
```
where `num` is a field we needed to add since our data does not contain any fields which we could assign to an Integer.

In terms of data, the fields are reasonably short (less than 12 characters) with the exception of col3, col5 and col6 which have the potential to get large (this is a rare occurance)

We have also noticed that upon loading Typesense it was taking approximately 5 minutes when there were 1.3 million records indexed.

Finally, here is the space usage

```
540M	db
486M	state
```

## Expected Behavior

Looking at the README, I was expecting the memory usage to be significantly less. However, this may just be a case of having many items being indexed/type of data.

## Actual Behavior

- 2.5 Million records indexed and are consuming approximately 1.8 GB of memory.
- 5 minute startup time when 1.3 million records indexed.

## Metadata

**Typsense Version**: 0.16.0 (official docker image)

**OS**: Debian GNU/Linux 9 (stretch)

**Cluster mode**: No - We are running standalone mode in docker with data persistence.

**Comment by kishorenc** [comment]
@eric-sciberras 👋 

Can you please post the output of the `/metrics.json` end-point? That will provide a break-up of the memory consumption. In general, though, the ballpark memory usage is 2x-3x the DB storage on disk. 

The growth will typically flatten over time since much of the overhead in indexing text comes from the finite number of words in the dataset which will not grow after a certain point. So, the next 2.5M records should take smaller than 1.8 GB of memory.

**Comment by eric-sciberras** [comment]
Hi @kishorenc :smile:, 

Thanks for the insight into how the memory usage scales.

Here is the output of `/metrics.json`
```
{
  "system_cpu1_active_percentage": "10.00",
  "system_cpu2_active_percentage": "10.00",
  "system_cpu_active_percentage": "5.26",
  "system_disk_total_bytes": "21003583488",
  "system_disk_used_bytes": "2879098880",
  "system_memory_total_bytes": "8176574464",
  "system_memory_used_bytes": "2874355712",
  "system_network_received_bytes": "335939370",
  "system_network_sent_bytes": "41495813",
  "typesense_memory_active_bytes": "1979891712",
  "typesense_memory_allocated_bytes": "1941451632",
  "typesense_memory_fragmentation_ratio": "0.02",
  "typesense_memory_mapped_bytes": "2053451776",
  "typesense_memory_metadata_bytes": "44080080",
  "typesense_memory_resident_bytes": "1979891712",
  "typesense_memory_retained_bytes": "265998336"
}
```
Note: We are currently building up the collection so these values correspond to 2.65 million entries.

**Comment by kishorenc** [comment]
The metrics look fine to me: very little fragmentation. The usage is along expected lines given the number of string fields being indexed, though I do have some items on my todo list to decrease the consumption further. So hopefully the memory footprint will grow smaller in future releases and will require nothing but a restart with a new binary. I recommend running in a 3-node configuration for rolling updates and high availability.

**Comment by kishorenc** [comment]
Closing this for now.

**Comment by eric-sciberras** [comment]
Hi, in the interest of sharing usage stats I just wanted to give an update.

Some details from the `/collections` endpoint (using the same schema defined above)
```
"num_documents":8588011,"num_memory_shards":4
```

 As for the `metrics.json` output
``` 
{
  "system_cpu1_active_percentage": "0.00",
  "system_cpu2_active_percentage": "0.00",
  "system_cpu_active_percentage": "4.76",
  "system_disk_total_bytes": "21003583488",
  "system_disk_used_bytes": "2276433920",
  "system_memory_total_bytes": "8176570368",
  "system_memory_used_bytes": "7059103744",
  "system_network_received_bytes": "400948826",
  "system_network_sent_bytes": "118131345",
  "typesense_memory_active_bytes": "5948129280",
  "typesense_memory_allocated_bytes": "5808136504",
  "typesense_memory_fragmentation_ratio": "0.02",
  "typesense_memory_mapped_bytes": "6139011072",
  "typesense_memory_metadata_bytes": "131326800",
  "typesense_memory_resident_bytes": "5948129280",
  "typesense_memory_retained_bytes": "1454776320"
}
```

**Comment by jasonbosco** [comment]
Thank you for sharing @eric-sciberras!

**Comment by surajthakur** [comment]
@jasonbosco  for the above case, does upgrading from 0.16.0 to 0.18.0 would help in less memory consumption ? 
Currently after typesense restart it took 30 minutes  to load `8588011`  documents.

Also if we move to a clustered (High Available) architecture, how would that help? 
Does the collections data gets divided among the three nodes or same data is replicated on all three nodes. In the latter case, the memory consumption would be the same as its is currently on single node. 

Could you suggest, if we should make any changes to our architecture. 

Thanks

**Comment by jasonbosco** [comment]
@surajthakur We haven't yet released the memory optimizations as of v0.18.0, they're still on our radar though. So upgrading to v0.18.0 will not change memory consumption rates. 

> Currently after typesense restart it took 30 minutes to load 8588011 documents.

Interesting. I've seen datasets with 32M rows load in 20 minutes on restart, from an SSD disk. Although the time it takes to reload the data depends on the size of each record. The reason it takes this long is because it's essentially re-indexing the data on restart. We had to do this to make upgrades between Typesense version as easy as restarting the process with a new version of the binary. So it's a trade-off between ease of upgrades vs restart times. 

One thing to remember is to only specify fields you want to search/filter/facet by in the collection's schema. You can have any number of additional unindexed fields in your records and you need not specify them in the schema. These unindexed fields will only be stored on disk. So if you have fields in the collections schema that you don't search/filter/facet by, if you remove them from the schema definition (not from the records), you'll see an improvement in memory usage.

> Also if we move to a clustered (High Available) architecture, how would that help?
> Does the collections data gets divided among the three nodes or same data is replicated on all three nodes. In the latter case, the memory consumption would be the same as its is currently on single node.

With multiple nodes in HA mode, the same data gets replicated to all nodes. Memory consumption and memory requirements do not change with HA. The reason to use an HA architecture is to ensure that one node going down doesn't affect the cluster's availability - it will still continue serving traffic.

**Comment by surajthakur** [comment]
Thanks for the insight @jasonbosco . We will play around the schema and see if we can filter any fields from indexing.
For HA, we are considering it, but we wanted to get the memory thing sorted out first. 
Will update here if we have any further queries
